File: LeetCode_Algorithm/125/125.js
var isPalindrome = function(s) {
  // sol1 84~86 79~76%
  // ascii code
  // var string = ""
  // for (let left=0; left < s.length; left++){
  //   // process.stdin.write(s[left].charCodeAt() + " ")
  //   let asciiCode = s[left].charCodeAt()

  //   if ( 65 <= asciiCode  && asciiCode <= 90){
  //     string += s[left].toLowerCase()
  //   }
  //   if (97 <= asciiCode && asciiCode <= 122){
  //     string += s[left]
  //   }
  //   if (48 <= asciiCode && asciiCode <= 57){
  //     string += s[left]
  //   }  
  // }
  
  // for (let left=0; left < parseInt(string.length/2); left++){
  //   if (string[left] != string[string.length-1-left]){
  //     return false
  //   }
  // }
  // return true


  // sol2 64~110 99~47%보다 빠름
  // var s = s.toLowerCase().split('')
  // var string = s.filter((chr)=>{
  //   if(isalnum(chr)){
  //     return true
  //   }
  // })
  // let right = string.length
  // for (let left=0; left < parseInt(string.length/2); left++){
  //   if (string[left] != string[right-left-1]){
  //     return false
  //   }
  // }
  // return true

  // sol3 89~99ms 72~61%보다 빠름
  // while을 사용해서 조금 더 깔끔함
  var left = 0
  var right = s.length-1

  while (left < right){
    while (left < right && !isalnum(s[left])){
      left += 1
    }
    while (left < right && !isalnum(s[right])){
      right -= 1
    }
    if (s[left].toLowerCase() != s[right].toLowerCase()){
      return false
    }
    left += 1
    right -= 1
  }

  return true

};
function isalpha(chr){
  return (((chr >= 'a') && (chr <= 'z')) || ((chr >= 'A') && (chr <= 'Z')))
}
function isdigit(chr){
  return ((chr >= '0') && (chr <= '9'))
}
function isalnum(chr){
  return ( isalpha(chr) || isdigit(chr) )
}

console.log(isPalindrome("A p P a"))

File: LeetCode_Algorithm/125/125.py
from http.client import CONTINUE
from re import I


class Solution:
    def isPalindrome(self, s: str) -> bool:
        
        # sol1 n 70~110ms 49~10% 보다 빠름
        # 아스키 코드로 바꾼 뒤 공백 등등 다 지우고 체크
        # string = ""
        # for c in s:
        #     if 65 <= ord(c) <= 90:
        #         string += chr(ord(c)+32)
        #     elif 97 <= ord(c) <= 122:
        #         string += c
        #     elif 48 <= ord(c) <= 57:
        #         string += c

        # j = len(string)-1
        # i = 0
        # while i <=j:
        #     if  string[i] != string[j]:
        #         return False
        #     i += 1
        #     j -= 1
        
        # sol2 n 70~110ms 49~10% 보다 빠름
        # i = 0
        # j = len(s)-1
        # chr_i, chr_j = "", ""
        # while i <= j:
        #     if 65 <= ord(s[i]) <= 90:
        #         chr_i = chr(ord(s[i])+32)
        #     elif 97 <= ord(s[i]) <= 122:
        #         chr_i = s[i]
        #     elif 48 <= ord(s[i]) <= 57:
        #         chr_i = s[i]
        #     else:
        #         i += 1; continue
            
        #     if 65 <= ord(s[j]) <= 90:
        #         chr_j = chr(ord(s[j])+32)
        #     elif 97 <= ord(s[j]) <= 122:
        #         chr_j = s[j]
        #     elif 48 <= ord(s[j]) <= 57:
        #         chr_j = s[j]
        #     else:
        #         j -= 1; continue
            
        #     if chr_i != "" and chr_j != "":
        #         if chr_i != chr_j:
        #             return False
        #         else:
        #             i += 1
        #             j -= 1

        # return True

        # sol3 60~70ms 50~42%보다 빠름
        # isalnum() 메소드를 통해 숫자나 알파벳을 확인
        # alnum_string = filter(lambda chr : chr.isalnum(), s)
        # lowercase_string = list(map(lambda chr : chr.lower(), alnum_string))
        # reverse_string = lowercase_string[::-1]
        # for i in range(len(lowercase_string)//2):
        #     if lowercase_string[i] != reverse_string[i]:
        #         return False
        # return True

        # sol4
        left = 0
        right = len(s)-1
        while left < right:
            while left < right and not s[left].isalnum():
                left += 1
            while left < right and not s[right].isalnum():
                right -= 1

            if s[left].lower() != s[right].lower():
                return False
            left += 1
            right -= 1

        return True



print(Solution().isPalindrome(".,"))